packages/apps/patterns/react-composer/src/yjs/space-provider.ts

```diff
@@ -0,0 +1,135 @@
+//
+// Copyright 2023 DXOS.org
+//
+
+import * as decoding from 'lib0/decoding';
+import * as encoding from 'lib0/encoding';
+import { Observable } from 'lib0/observable';
+import * as awarenessProtocol from 'y-protocols/awareness';
+import { Doc } from 'yjs';
+
+import { GossipMessage, Space } from '@dxos/client';
+import { log } from '@dxos/log';
+
+type Awareness = awarenessProtocol.Awareness;
+
+// Based on https://github.com/yjs/y-webrtc/blob/88baab2/src/y-webrtc.js.
+
+const messageAwareness = 1;
+const messageQueryAwareness = 3;
+
+/**
+ * Yjs awareness provider on top of a DXOS space.
+ */
+export class SpaceProvider extends Observable<any> {
+  private readonly _space: Space;
+  private readonly _awareness: Awareness;
+  private readonly _clientId: number;
+  private readonly _channel: string;
+
+  constructor({
+    space,
+    doc,
+    channel = 'dxos.composer.status',
+    awareness
+  }: {
+    space: Space;
+    doc: Doc;
+    channel?: string;
+    awareness?: Awareness;
+  }) {
+    super();
+    this._space = space;
+    this._awareness = awareness ?? new awarenessProtocol.Awareness(doc);
+    this._channel = channel;
+    this._clientId = doc.clientID;
+
+    this._awareness.on('update', this._handleAwarenessUpdate.bind(this));
+    this._space.listen(this._channel, this._handleSpaceMessage.bind(this));
+
+    if (typeof window !== 'undefined') {
+      window.addEventListener('beforeunload', this._handleBeforeUnload.bind(this));
+    } else if (typeof process !== 'undefined') {
+      process.on('exit', this._handleBeforeUnload.bind(this));
+    }
+
+    // Post queryAwareness.
+    const encoderAwarenessQuery = encoding.createEncoder();
+    encoding.writeVarUint(encoderAwarenessQuery, messageQueryAwareness);
+    void this._space.postMessage(this._channel, encoding.toUint8Array(encoderAwarenessQuery));
+
+    // Post local awareness state.
+    const encoderAwarenessState = encoding.createEncoder();
+    encoding.writeVarUint(encoderAwarenessState, messageAwareness);
+    encoding.writeVarUint8Array(
+      encoderAwarenessState,
+      awarenessProtocol.encodeAwarenessUpdate(this.awareness, [this._clientId])
+    );
+    void this._space.postMessage(this._channel, encoding.toUint8Array(encoderAwarenessState));
+  }
+
+  get awareness(): Awareness {
+    return this._awareness;
+  }
+
+  private _handleAwarenessUpdate({ added, updated, removed }: any, origin: any) {
```

Comment:
awareness vs. presence?

Reply:
the awareness name comes from yjs, so we're bound to that